### DISCUSSION TITLE
Mounting a Host Directory in a Podman Container While Preserving Container’s Directory Content

### DISCUSSION THREAD

**Discussion by honhimW** [containers/podman#20460] [Q&A] [answered]
Hello,

I have a question regarding the use of Podman. I am trying to mount a host directory to a directory inside a Podman container while preserving the content that is already inside the container’s directory.
> **Because sometimes the directory in the container where the content needs to be persisted already contains some data required for initialization and data generated during runtime.**

I understand that I can use the `-v` or `--volume` option when running a container to mount a host directory to a container directory. The format for this option is `-v <host-dir>:<container-dir>`. However, this operation overwrites the content inside the container’s directory.

Is there an optional parameter that supports creating a directory directly when running a container? If not, similar work needs to be done through the following steps:

1. Create the directory on the host that you want to mount, e.g., `/data/<container_name>`;
2. Run a new container, e.g., `podman run -it --name <container_name> <image>`;
3. Copy files from the new container to the host’s directory: `podman cp <container_name>:/path/to/container/dir /data/<conatiner_name>`;
4. Stop and remove the temporary container;
5. Run a new container and mount the host’s directory: `podman run -it -v /data/<container_name>:/path/to/container/dir`;

**Comment by carrulop** [comment]
Hello, I am interested in this feature? Is it implemented in podman?

Thanks!

**Comment by rhatdan** [comment]
I really do not understand what you are after here?

Do you want to volume mount into a container on a directory, that still has the contents of the underlying directory?

Something similar to this exists for --tmpfs 

              • tmpcopyup: Enable copyup from the image directory at the same location to the tmpfs/ramfs. Used by default.

**Reply by honhimW** [reply]
> No! Is not about the `--tmpfs` option😡

I want to persist the container’s contents to the host, including both the initial data from the container image and new data generated at runtime.(As I said at the beginning).

For example, I need to run a MySQL container and persist the contents of the `/var/lib/mysql` directory. This directory already contains essential files from the image needed for startup, as well as runtime-generated database data. I want to persist these to a directory on the host.

When using Docker, if you use `docker run -v` to mount a non-existent host directory, it will automatically create and fill it with the container’s data. However, if the host directory already exists, it will overwrite the container’s contents.
With Podman, if the corresponding directory doesn’t exist on the host, the container won’t start at all. Therefore, I have to go through the 5 steps mentioned above to achieve this functionality, which is very frustrating and requires creating a temporary container.

**Reply by rhatdan** [reply, marked as answer]
Interesting I was not aware of this behaviour.

This seems similar to the behavior of using a named volume.

If you do a `podman run -v mysql:/var/lib/mysql ...` Does it do what you expect, other then not using a directory on disk?

**Reply by honhimW** [reply]
Well, it's indeed a solution. Now the operating steps become:

### 1. optional, modify storage.conf if necessary.
[storage.conf](https://github.com/containers/podman/blob/main/vendor/github.com/containers/storage/storage.conf)
- /usr/containers/storage.conf
- /etc/containers/storage.conf
- $HOME/.config/containers/storage.conf
```properties
rootless_storage_path = "$HOME/.local/share/containers/storage"
```
```shell
# restart podman
systemctl --user restart podman
```

### 2. create a named volume
```shell
podman volume create {volume_name}
```
### 3. run the container with volume
```shell
podman run -v {volume_name}:/path/to/be/persisted ...
```
### get directory
```shell
# can't use remote client
/usr/bin/podman unshare /usr/bin/podman volume mount {volume_name}
# echo
# ~/.local/share/containers/storage/volumes/{volume_name}/_data
```
